### DISCUSSION TITLE
cant send ether to function with arguments

### DISCUSSION THREAD

**Discussion by vishwa7dcap786** [Cyfrin/foundry-full-course-cu#668] [Q&A]
contract--
function buyTicket(uint256 _lotteryId, uint16[4] calldata TicketNo) external payable nonReentrant {
        if (lotteries[_lotteryId].status != Status.Open) {
            revert Lottery_closed();
        }
        uint256 _TicketBuyingPrice = lotteries[_lotteryId].TicketBuyingPrice;
        if (msg.value != _TicketBuyingPrice * TicketNo.length) {
            revert SendCorrectTicketPrice();
        }
        require(TicketNo.length != 0, "No ticket specified");
        TotalTicketsForParticipant[_lotteryId][msg.sender] += TicketNo.length;
        console.log(msg.value);

        for (uint256 i = 0; i < TicketNo.length; i++) {
            uint256 thisTicketNumber = TicketNo[i];
            uint8 Digits = numDigits(thisTicketNumber);
            require(Digits == 4, "only four digit numbers are allowed");
            Tickets[currentTicketId] = Ticket({number: thisTicketNumber, owner: msg.sender});

            ParticipantsTicketIdInLottery[_lotteryId][msg.sender].push(currentTicketId);
            emit TicketsPurchase(currentTicketId, msg.sender, _lotteryId, thisTicketNumber);
            currentTicketId++;
        }
        lotteries[_lotteryId].amountCollectedInLottery += msg.value;
    }
-----------------------------------------------------------------------------------------------------------------------------------------------------------
tests--
// SPDX-Lisence-Identifier:MIT

pragma solidity 0.8.20;

import {Test, console} from "forge-std/Test.sol";
import {DeployLottery} from "../script/DeployLottery.s.sol";
import {Lottery} from "../src/lot.sol";

contract LotteryTest is Test {
    DeployLottery Deployer;
    Lottery lottery;
    address payable admin = payable(address(0x5B38Da6a701c568545dCfcB03FcB875f56beddC4));
    address operator = 0xAb8483F64d9C6d1EcF9b849Ae677dD3315835cb2;
    address payable treasury = payable(address(0x4B20993Bc481177ec7E8f571ceCaE8A9e22C02db));

    address user1 = 0x17F6AD8Ef982297579C203069C1DbfFE4348c372;
    address user2 = vm.addr(234);
    address user3 = vm.addr(345);

    function setUp() public {
        Deployer = new DeployLottery();
        (lottery) = Deployer.run();
        vm.deal(user1, 10 ether);
        vm.deal(user2, 10 ether);
        vm.deal(user3, 10 ether);
        vm.deal(operator, 10 ether);
    }
function testBuyTickets() public {
        vm.startPrank(user1);

        uint16[4] memory arg = [1122, 2233, 3344, 4455];

        lottery.buyTicket{value: 4 ether}(1, arg);
       
        vm.stopPrank();

        assertEq(address(lottery).balance, 4 ether);
    }
------------------------------------------------------------------------------------------------------------------------------------------------------------------
error--

[FAIL. Reason: SendCorrectTicketPrice()] testBuyTickets() (gas: 27446)
Traces:
  [2630758] LotteryTest::setUp() 
    ├─ [1332030] → new DeployLottery@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← 6211 bytes of code
    ├─ [1210388] DeployLottery::run() 
    │   ├─ [0] VM::startBroadcast() 
    │   │   └─ ← ()
    │   ├─ [1173286] → new Lottery@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   └─ ← 5416 bytes of code
    │   ├─ [0] VM::stopBroadcast() 
    │   │   └─ ← ()
    │   └─ ← Lottery: [0x90193C961A926261B756D1E5bb255e67ff9498A1]
    ├─ [0] VM::deal(0x17F6AD8Ef982297579C203069C1DbfFE4348c372, 10000000000000000000 [1e19]) 
    │   └─ ← ()
    ├─ [0] VM::deal(0x4F81e991f76276A17cA92a1321f37189b1727F77, 10000000000000000000 [1e19]) 
    │   └─ ← ()
    ├─ [0] VM::deal(0xc0981804421D1C21f4B77eF722329C47d45D5A31, 10000000000000000000 [1e19]) 
    │   └─ ← ()
    ├─ [0] VM::deal(0xAb8483F64d9C6d1EcF9b849Ae677dD3315835cb2, 10000000000000000000 [1e19]) 
    │   └─ ← ()
    └─ ← ()

  [27446] LotteryTest::testBuyTickets() 
    ├─ [0] VM::startPrank(0x17F6AD8Ef982297579C203069C1DbfFE4348c372) 
    │   └─ ← ()
    ├─ [9919] Lottery::buyTicket{value: 4000000000000000000}(1, [1122, 2233, 3344, 4455]) 
    │   └─ ← "SendCorrectTicketPrice()"
    └─ ← "SendCorrectTicketPrice()"

Test result: FAILED. 0 passed; 1 failed; finished in 1.67ms

Failing tests:
Encountered 1 failing test in test/LotteryTest.t.sol:LotteryTest
[FAIL. Reason: SendCorrectTicketPrice()] testBuyTickets() (gas: 27446)
  
i have set the ticket price to 1 ether. so buying four Tickets for four ether.
still contract  is not receiving any amount . but function that has no args receives ether.

how to fix it?

**Comment by theirrationalone** [comment]
Let's say, For Example:  If we have `ticketBuingPrice` = 1 ether, on lottery at `lotteryId` = 1, and we have `ticketsLength` == `TicketNo.length` = 4.

Now, We're sending `4 ether`, then msg.value = 4 ether, `ticketBuingPrice` = 1 ether, `TicketNo.Length` = 4.

calcs: 4 ether == 4 ether // (1 ether * 4)

```solidity
uint256 _TicketBuyingPrice = lotteries[_lotteryId].TicketBuyingPrice;
if (msg.value != _TicketBuyingPrice * TicketNo.length) {
revert SendCorrectTicketPrice();
}
```

Reason: You're not setting `_TicketBuyingPrice` equal to 1 ether, of the lottery with `_lotteryId` = 1, Or If you're sure then Please provide full contract code.

**Reply by vishwa7dcap786** [reply]
// SPDX-Lisence-Identifier:MIT

pragma solidity ^0.8.19;

import "@openzeppelin-contracts/security/ReentrancyGuard.sol";
import {console} from "forge-std/console.sol";

contract Lottery is ReentrancyGuard {
    error SendCorrectTicketPrice();
    error WinnersLength_CantBeZero();
    error Lottery_closed();
    error Lottery_PriceIsNotClaimable();
    error TicketBuyingPriceCantBeZero();
    error Lottery_YouAreNotTicketOwner();
    error LotteryIsNotOpen_CantSetPrice();
    error Lottery_TransactionNotSuccess();

    event TicketsPurchase(
        uint256 TicketId, address indexed participant, uint256 indexed lotteryId, uint256 numberTicket
    );
    event PriceClaimed(
        uint256 TicketId, address participant, uint256 lotteryId, uint256 TicketNumber, uint256 PriceClaimed
    );

    address payable public treasury;
    address payable public admin;
    address public operator;
    uint256 private constant PRECISION = 100;
    uint256 private currentLotteryNo;
    uint256 private currentTicketId;

    enum Status {
        Open,
        Close,
        Claimable
    }

    struct lottery {
        Status status;
        uint256 startTime;
        uint256 endTime;
        uint256 TicketBuyingPrice;
        uint256 firstTicketId;
        uint256 firstTicketIdNextLottery;
        uint256 amountCollectedInLottery;
        uint256 AdminPercent;
        uint32 finalNumber;
    }

    struct Ticket {
        uint256 number;
        address owner;
    }

    struct winnerId {
        uint256 TicketId;
        uint256 PositionNo;
        uint256 priceAmount;
    }

    mapping(uint256 Ticketid => Ticket) private Tickets;
    mapping(uint256 lotteryId => lottery) private lotteries;
    mapping(uint256 LotteryNo => mapping(address participants => uint256[] TicketNo)) public
        ParticipantsTicketIdInLottery;
    mapping(uint256 LotteryNo => mapping(uint256 TicketId => winnerId)) public LotteryWinners;
    mapping(uint256 LotteryNo => mapping(address participant => uint256 TotalTickets)) public TotalTicketsForParticipant;

    function viewTickets(uint256 TicketId) public view returns (uint256 ticketnumber, address ticketowner) {
        ticketnumber = Tickets[TicketId].number;
        ticketowner = Tickets[TicketId].owner;
    }

    constructor(address payable _treasury, address payable _admin, address _operator) {
        treasury = _treasury;
        admin = _admin;
        operator = _operator;
    }

    modifier onlyAdmin() {
        require(msg.sender == operator, "Access denied, Admin can only access this funtion");
        _;
    }

    function buyTicket(uint256 _lotteryId, uint16[4] calldata TicketNo) external payable nonReentrant {
        if (lotteries[_lotteryId].status != Status.Open) {
            revert Lottery_closed();
        }
        uint256 _TicketBuyingPrice = lotteries[_lotteryId].TicketBuyingPrice;
        if (msg.value != _TicketBuyingPrice * TicketNo.length) {
            revert SendCorrectTicketPrice();
        }
        require(TicketNo.length != 0, "No ticket specified");
        TotalTicketsForParticipant[_lotteryId][msg.sender] += TicketNo.length;
        console.log(msg.value);

        for (uint256 i = 0; i < TicketNo.length; i++) {
            uint256 thisTicketNumber = TicketNo[i];
            uint8 Digits = numDigits(thisTicketNumber);
            require(Digits == 4, "only four digit numbers are allowed");
            Tickets[currentTicketId] = Ticket({number: thisTicketNumber, owner: msg.sender});

            ParticipantsTicketIdInLottery[_lotteryId][msg.sender].push(currentTicketId);
            emit TicketsPurchase(currentTicketId, msg.sender, _lotteryId, thisTicketNumber);
            currentTicketId++;
        }
        lotteries[_lotteryId].amountCollectedInLottery += msg.value;
    }

    function SetWinners(uint256 lotteryNO, uint64[3][3] memory winners) external nonReentrant onlyAdmin {
        require((lotteries[lotteryNO].status == Status.Close), "Not time to set lottery Winners");
        if (winners.length == 0) {
            revert WinnersLength_CantBeZero();
        }
        for (uint256 i = 0; i < winners.length; i++) {
            uint64[3] memory winner = winners[i];
            console.log(winner[0], winner[1], winner[2]);
            LotteryWinners[lotteryNO][winner[0]].TicketId = winner[0];
            LotteryWinners[lotteryNO][winner[0]].PositionNo = winner[1];
            LotteryWinners[lotteryNO][winner[0]].priceAmount = winner[2];
        }
        lotteries[lotteryNO].status = Status.Claimable;
    }

    function setTicketPrice(uint256 _lotteryNo, uint256 _TicketBuyingPrice) external nonReentrant onlyAdmin {
        if (lotteries[_lotteryNo].status != Status.Open) {
            revert LotteryIsNotOpen_CantSetPrice();
        }
        if (_TicketBuyingPrice == 0) {
            revert TicketBuyingPriceCantBeZero();
        }
        lotteries[_lotteryNo].TicketBuyingPrice = _TicketBuyingPrice;
    }

    function operatorAccess() public view returns (bool) {
        if (msg.sender == operator) {
            return true;
        } else {
            return false;
        }
    }

    function checkAddress(address addr1) public view returns (bool) {
        if (addr1 == operator) {
            return true;
        } else {
            return false;
        }
    }

    function claimAmount(uint256 lotteryNo) external payable nonReentrant {
        address sender = msg.sender;
        if (lotteries[lotteryNo].status != Status.Claimable) {
            revert Lottery_PriceIsNotClaimable();
        }
        //uint256[] memory UserTicketIds;

        uint256 TotalTickets = TotalTicketsForParticipant[lotteryNo][sender];
        // for(uint256 i = 0 ; i < TotalTickets; i++){
        // ParticipantsTicketIdInLottery[lotteryNo][msg.sender][i]
        // }
        uint256 LotterypriceAmount;

        for (uint256 i = 0; i < TotalTickets; i++) {
            uint256 TicketID = ParticipantsTicketIdInLottery[lotteryNo][sender][i];
            console.log(TicketID);
            LotterypriceAmount = LotteryWinners[lotteryNo][TicketID].priceAmount;
            console.log(LotterypriceAmount);
            console.log(Tickets[TicketID].owner);
            // if (msg.sender != Tickets[TicketId].owner) {
            //     revert Lottery_YouAreNotTicketOwner();
            // }
            emit PriceClaimed(
                TicketID, sender, lotteryNo, Tickets[TicketID].number, LotteryWinners[lotteryNo][TicketID].priceAmount
            );
            if (LotterypriceAmount != 0) {
                bool success = payable(sender).send(LotterypriceAmount);
                if (success == false) {
                    revert Lottery_TransactionNotSuccess();
                }
            }
            Tickets[TicketID].owner = address(0);
        }
        //uint256 TicketNumber = Tickets[TicketId].number;
    }

    function setLotteryOpen(uint256 _endTime, uint256 _TicketBuyingPrice, uint256 _AdminPercent) external onlyAdmin {
        require(
            (currentLotteryNo == 0) || (lotteries[currentLotteryNo].status == Status.Claimable),
            "Not time to start lottery"
        );
        currentLotteryNo++;
        lotteries[currentLotteryNo] = lottery({
            status: Status.Open,
            startTime: block.timestamp,
            endTime: _endTime,
            TicketBuyingPrice: _TicketBuyingPrice,
            firstTicketId: currentTicketId,
            firstTicketIdNextLottery: currentTicketId,
            finalNumber: 0,
            amountCollectedInLottery: 0,
            AdminPercent: _AdminPercent
        });
    }

    function setLotteryClosed(uint256 _lotteryNo) external onlyAdmin {
        require(lotteries[_lotteryNo].status == Status.Open, "Lottery not open");
        lotteries[_lotteryNo].firstTicketIdNextLottery = currentTicketId;

        lotteries[_lotteryNo].status = Status.Close;

        moveToAdmin();
    }

    function moveToAdmin() private {
        uint256 balance = lotteries[currentLotteryNo].amountCollectedInLottery;
        uint256 adminPercent = lotteries[currentLotteryNo].AdminPercent;
        uint256 adminAmount = (balance * adminPercent) / PRECISION;
        bool success = payable(admin).send(adminAmount);
        require(success, "Send failed");
    }

    function viewLottery(uint256 _lotteryId) external view returns (lottery memory) {
        return lotteries[_lotteryId];
    }

    function viewCurrentLotteryId() external view returns (uint256) {
        return currentLotteryNo;
    }

    function numDigits(uint256 number) private pure returns (uint8) {
        uint8 digits = 0;
        while (number != 0) {
            number /= 10;
            digits++;
        }
        return digits;
    }

    receive() external payable {}

    /* function deposit() public payable {
        contractBalance += msg.value;
    }

    function withdraw(uint256 amount) public {
    uint256 currentBalance = address(this).balance;
    require(amount <= currentBalance, "Insufficient balance in contract.");
     payable(msg.sender).transfer(amount);
    }*/
}

**Reply by theirrationalone** [reply]
Hey, @vishwa7dcap786 ! Please execute this function `setTicketPrice` before Buying Ticket!

Do something like this...

```solidity
function testBuyTickets() public {
vm.startPrank(user1);

    uint16[4] memory arg = [1122, 2233, 3344, 4455];
    setTicketPrice(1, 1 ether);    // here 😉, you need to implement this line.
    lottery.buyTicket{value: 4 ether}(1, arg);
   
    vm.stopPrank();

    assertEq(address(lottery).balance, 4 ether);
}
```

**Comment by vishwa7dcap786** [comment]
TicketPrice will be updated when opening the lottery  using the setLotteryOpen function

**Comment by vishwa7dcap786** [comment]
contract is not receiving any ethers when sending ethers to the function with arguments

**Comment by theirrationalone** [comment]
In your traces logs, there's no footprint of `setLotteryOpen` function. 

Still you can set ticket price using this function: `setTicketPrice(1, 1 ether);`

You should have an `admin` address to make this happen... There you can have a modifier that can set ticket price in you test. 

Something like this:

```solidity
modifier setTckPrice() {
    address admin = lottery.getAdmin();
    vm.prank(admin);
    setTicketPrice(1, 1 ether);
}

function testBuyTickets() public setTckPrice {
    vm.startPrank(user1);
    uint16[4] memory arg = [1122, 2233, 3344, 4455];
    lottery.buyTicket{value: 4 ether}(1, arg);
    vm.stopPrank();

    assertEq(address(lottery).balance, 4 ether);
}
```

**Reply by vishwa7dcap786** [reply]
there is a function like this in the above contract

 function setLotteryOpen(uint256 _endTime, uint256 _TicketBuyingPrice, uint256 _AdminPercent) external onlyAdmin {
    require(
        (currentLotteryNo == 0) || (lotteries[currentLotteryNo].status == Status.Claimable),
        "Not time to start lottery"
    );
    currentLotteryNo++;
    lotteries[currentLotteryNo] = lottery({
        status: Status.Open,
        startTime: block.timestamp,
        endTime: _endTime,
        TicketBuyingPrice: _TicketBuyingPrice,
        firstTicketId: currentTicketId,
        firstTicketIdNextLottery: currentTicketId,
        finalNumber: 0,
        amountCollectedInLottery: 0,
        AdminPercent: _AdminPercent
    });
}

**Reply by theirrationalone** [reply]
Yess exactly, You only need an Admin to do this.... 
Make a helper function to get admin into your lottery contract and make a prank with it, like this... 

```solidity
function testBuyTickets() public {
    address admin = lottery.getAdmin();

    vm.startPrank(admin);
    setLotteryOpen(_endTime, _TicketBuyingPrice, _AdminPercent);
    uint16[4] memory arg = [1122, 2233, 3344, 4455];
    lottery.buyTicket{value: 4 ether}(1, arg);
    vm.stopPrank();

    assertEq(address(lottery).balance, 4 ether);
}

```

Or you can `transferOwnership` to user1 to make him/her Admin. 

If this doesn't help you then Please Provide Your lottery Repo link... So i'll be able to test Your repo...!

**Reply by theirrationalone** [reply]
And if you are sure with it... Then please also provide `deployscript` code...
